File: hashmap_doubleHashing.py
# Time Complexity : put() - O(1), get() - O(1) 
# Space Complexity : O(m^2) - m=bucketsize
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

'''
I did the problem using double hashing similar to hasset, except for the fact that this required to explicity store key-value
pairs as tuple.
'''

class MyHashMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.bucket_size = 1000
        self.bucketItem_size = 1000 + 1
        self.buckets = [None] * self.bucket_size
        

    def put(self, key: int, value: int) -> None:
        """
        value will always be non-negative.
        """
        hashed_key = key % self.bucket_size
        hashedItem_key = int(key // self.bucketItem_size)
        
        if self.buckets[hashed_key] is None:
            self.buckets[hashed_key] = [(None,None)] * self.bucketItem_size
            
        self.buckets[hashed_key][hashedItem_key] = (key, value)
            
        

    def get(self, key: int) -> int:
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        """
        hashed_key = key % self.bucket_size
        hashedItem_key = int(key // self.bucketItem_size)
        
        if self.buckets[hashed_key] is not None:
            k,v = self.buckets[hashed_key][hashedItem_key]
            if k is not None and v is not None:
                return v
                         
        return -1
        

    def remove(self, key: int) -> None:
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        """
        hashed_key = key % self.bucket_size
        hashedItem_key = int(key // self.bucketItem_size)
        if self.buckets[hashed_key] is not None:
            self.buckets[hashed_key][hashedItem_key] = (None, None)
        


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)

File: HashMap-DoubleHashing.java
//Exercise_1 : Design Hashmap
// Time Complexity : O(1)  -> for all operations
// Space Complexity : O(N+K)=> Total Elements where N is size of parent array and K is the size of child list where value is stored(Linked List)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :  No


// Your code here along with comments explaining your approach

//Note: Here Double Hashing is implemented. I created a class Node and initialised 2d array of Nodes. For each index I'm storing 1D array of nodes and then storing values on the index of 1D Array 


class MyHashMapDoubleHashing {

    
    //Using Double Hashing
    private int bucketSize;
    private int bucketItemsSize;
    
    class Node{
        private int val;
        
        Node(int val){
            this.val = val;
        }
    }
    
    private int getBucketHash(int key){
        return key%bucketSize;
    }
    
    private int getBucketItemHash(int key){
        return key/bucketItemsSize;
    }
    
    private Node[][] nodes;
    
    /** Initialize your data structure here. */
    public MyHashMapDoubleHashing() {
        bucketSize = 1000;
        bucketItemsSize = 1001;
        nodes = new Node[bucketSize][];
    }
    
    /** value will always be non-negative. */
    public void put(int key, int value) {
        
        int bucketIndex = getBucketHash(key);
        int bucketItemIndex = getBucketItemHash(key);
        
        if(nodes[bucketIndex]==null){
            nodes[bucketIndex] = new Node[bucketItemsSize];
        }
        
        if(nodes[bucketIndex][bucketItemIndex]==null){
            nodes[bucketIndex][bucketItemIndex] = new Node(value);
        }else{
           Node node = nodes[bucketIndex][bucketItemIndex];
           node.val = value; 
        }
        
        
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int key) {
        
        int bucketIndex = getBucketHash(key);
        int bucketItemIndex = getBucketItemHash(key);
        
        if(nodes[bucketIndex]==null){
            return -1;
        }
        
        if(nodes[bucketIndex][bucketItemIndex]==null){
            return -1;
        }else{
           Node node = nodes[bucketIndex][bucketItemIndex];
           return node.val; 
        }
        
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int key) {
         
        int bucketIndex = getBucketHash(key);
        int bucketItemIndex = getBucketItemHash(key);
        
        if(nodes[bucketIndex]==null){
            return;
        }
        
        if(nodes[bucketIndex][bucketItemIndex]==null){
            return;
        }else{
           nodes[bucketIndex][bucketItemIndex] = null;
        }
        
    }
}